### PR TITLE
Notebook for Mixture Density Network

### PR DESCRIPTION
**Pull request by vdutor**

### CODE REVIEW

File: doc/source/notebooks/manual/tailor/mixture_density_network.ipynb [hunk 1]
```diff
@@ -4,15 +4,745 @@
...
+    "\n",
+    "Mixture density networks (MDNs) are a parametric class of models that allow for conditional density estimation. They are built out of 2 parts: a neural net and a Mixture of Gaussians (MoG). The neural net is  responsible for producing the characteristics of the MoG. In practice, given that the MoG consists of $M$ Gaussians, the neural net will output a collection of $M$ means, variances and weights $\\{\\mu_m, \\sigma_m^2, \\pi_m\\}_{m=1}^M$. The produced means, variances and weights are used to define the conditional probability distribution function \n",
+    "$$\n",
+    "p(Y = y\\,|\\,X = x) = \\sum_{m=1}^{M} \\pi_{m}(x)\\,\\mathcal{N}\\big(y\\, \\left|\\,\\mu_{m}(x), \\sigma_{m}^2(x)\\big)\\right.\n",
+    "$$\n",
+    "Each of the parameters  $\\pi_{m}(x), \\mu_{m}(x), \\sigma_{m}(x)$ of the distribution will be determined by the neural network, as a function of the input $x$.\n",
+    "\n",
+    "We train the MDN's neural net by optimizing the model's likelihood\n",
+    "$$\n",
+    " \\mathcal{L} \\triangleq \\operatorname{argmax}_{\\Theta} \\prod_{n=1}^N p(Y = y_n\\,|\\,X = x_n),\n",
```
Comment (awav) [inline doc/source/notebooks/manual/tailor/mixture_density_network.ipynb]: It is not rendered!

File: doc/source/notebooks/manual/tailor/mixture_density_network.ipynb [hunk 2]
```diff
@@ -4,15 +4,745 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class MDN(Model):\n",
+    "    \n",
+    "    def __init__(self, X, Y, inner_dims=[10, 10,], activation=tf.nn.tanh, num_mixtures=5):\n",
+    "        Model.__init__(self)\n",
+    "        \n",
+    "        self.Din = X.shape[1]\n",
+    "        self.dims = [self.Din, ] + list(inner_dims) + [3 * num_mixtures]\n",
```
Comment (awav) [inline doc/source/notebooks/manual/tailor/mixture_density_network.ipynb]: Can you clarify what happens in this line? A comment or an explanation in the cell would be good.

File: doc/source/notebooks/manual/tailor/mixture_density_network.ipynb [hunk 3]
```diff
@@ -4,15 +4,745 @@
...
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import gpflow\n",
+    "from gpflow.models.model import Model\n",
+    "from gpflow.params.dataholders import Minibatch, DataHolder\n",
+    "from gpflow.params import Parameter, ParamList\n",
+    "from gpflow.training import AdamOptimizer, ScipyOptimizer\n",
+    "from gpflow.decors import params_as_tensors, autoflow\n",
```
Comment (awav) [inline doc/source/notebooks/manual/tailor/mixture_density_network.ipynb]: `params_as_tensors` and `autoflow` is available though `gpflow`, pls, do not use `decors`

File: doc/source/notebooks/manual/tailor/mixture_density_network.ipynb [hunk 4]
```diff
@@ -4,15 +4,745 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "CM = plt.get_cmap('Blues')\n",
+    "N_plot = 100\n",
+    "\n",
+    "def make_grid(xx, yy):\n",
```
Comment (awav) [inline doc/source/notebooks/manual/tailor/mixture_density_network.ipynb]: Would be nice if you could hide this methods in aux `.py` file.

File: doc/source/notebooks/manual/tailor/mixture_density_network.ipynb [hunk 5]
```diff
@@ -4,15 +4,745 @@
...
...YltdCRWpoMIpkQCXV9qmCnmcMRMUEEBNAJIqZcX+UJJ4N4IlnurNoGayslhssWFcdgfoZalxAnB8rkdWIeOU79xilR7qOP+ej4ShVOGx3DKOcdlnFPTGJaUNPHqSF/AM0T4jxBPtUD7UIEeH6t9ELI+4zbfpr/rs7WUwp/NlpEHVPnFVcor+/+6SYf/vbjXLa6xvUbR457+4Yb8L+++jBbl1eoFExk8gBHyjY7ckNz64eLmYpeK5jZ8x8pOfjD6gdTTZ+jdRdQhPO2Ryd5ztYxTF1wxZqBTAG//9Acm8cqC1pN/FCy83e+xurBIl9/701sGjv+8FzB1E97kA960y4MIbLGwKNzHWxTP6YspI+zh6dDksX7Pnsfc52Av3/zVSf9O31SvIRhGoJBw6JkS5puSMePKFgRHU+nFUR4UUQ7VPmZnTCk7oUEMqblSSZaIW0/yhTcgmVQsAxCKSlYNVbN7KEceoz6LtXVK3AKDtboFsyN6zGTjnvHVHaMNPUi1X+EFqPqTGVSdyqQUhF1cssvUqpcVxlDJwppJVmfVctkZTVSSnFizxh2bIZLFhXH7FEv0ot5MjyfGHeV4mTJVPQJcR/Hwg+VMuwmZDgIe8mwlBBKmanCGXInZP331aljIYIsY9DiOFl9OrUDbV5Bzv+tTkSQW16IY+pJpXx3iC0TGM4BNoyW+c6v3pL5iV/ywW/z/pdd0tOY9293H+RTdz6BlPDdX7uFd9+yidGqzbrhEl+69zChjHnZtpU9r0HB0jNCXyuamTJWK5hsToimhopSS09IJpoea/wiZVvjV/7lB7zlWRt49Y7VTLf8HpINapB557ohVg44J3x+w2Wb4bK9KK/VWLV7X45lYCeEfaqpVhn7BPnc4diVIIWlqCB/5f4jlG2DZ20a4d3P3ZxljJ8s+qT4KQDLEAyVraySuWmE6K6GHwpCGWexZlPtkNlOxFwnzIov0oN4mLTqFUydNZHPsoJOeXiAtT+4A7FiBWa1AjfdgLl8SCnUQnSjzXKZmzKnOqtIfgCJlFpm5YhipSR7UYQXSWY9n6mOz2wnomLrjJUsRoo2uqZRMnVqlkWlYFApmFkOppZ97WZHpMpwnqBkQ3eilyD30UeKNFqtE0R4gcxWN1LfcBBJtaqRswB17ToagpxHvf/eOiMslEGcL+k4VYI8X0H2w4iCZSBlTN0NMvKUnt
```
Comment (awav) [inline doc/source/notebooks/manual/tailor/mixture_density_network.ipynb]: You can remove this section, it is covered by GPflows' documentation.

File: doc/source/notebooks/manual/tailor/mixture_density_network.ipynb [hunk 6]
```diff
@@ -4,15 +4,453 @@
...
+    "\n",
+    "Mixture density networks (MDNs) are a parametric class of models that allow for conditional density estimation. They are built out of 2 parts: a neural net and a Mixture of Gaussians (MoG). The neural net is  responsible for producing the characteristics of the MoG. In practice, given that the MoG consists of $M$ Gaussians, the neural net will output a collection of $M$ means, variances and weights $\\{\\mu_m, \\sigma_m^2, \\pi_m\\}_{m=1}^M$. The produced means, variances and weights are used to define the conditional probability distribution function \n",
+    "$$\n",
+    "p(Y = y\\,|\\,X = x) = \\sum_{m=1}^{M} \\pi_{m}(x)\\,\\mathcal{N}\\big(y\\, \\left|\\,\\mu_{m}(x), \\sigma_{m}^2(x)\\big)\\right.\n",
+    "$$\n",
+    "Each of the parameters  $\\pi_{m}(x), \\mu_{m}(x), \\sigma_{m}(x)$ of the distribution will be determined by the neural network, as a function of the input $x$.\n",
+    "\n",
+    "We train the MDN's neural net by optimizing the model's likelihood\n",
+    "$$\n",
+    " \\mathcal{L} \\triangleq \\text{argmax}_{\\Theta} \\prod_{n=1}^N p(Y = y_n~|~X = x_n),\n",
```
Comment (vdutor) [inline doc/source/notebooks/manual/tailor/mixture_density_network.ipynb]:
```suggestion
    " \\mathcal{L} \\triangleq \\text{argmax}_{\\Theta} \\prod_{n=1}^N p(Y = y_n | X = x_n),\n",
```

**Pull request merged**